bcgov/cas-registration
# 🏗️ Nextauth JWT & DB user role protected routes

Addresses [Ticket 265](https://app.zenhub.com/workspaces/team-moose-sprint-board-60d216175a2689000ec774d0/issues/gh/bcgov/cas-registration/265)

## 🚧 Changes
    - Modify client/app/api/auth/[...nextauth]/route.ts
    - Modify client/middlewares/withAuthorization.tsx
    - Modify client/app/middleware.ts
    - Modify client/app/(authenticated)/dashboard
    - Modify client/app/data/dashboard
    - Add protected routes client/app/(authenticated)/:
        - bceidbusiness/industry_user
        - bceidbusiness/industry_user_admin
        - idir/cas_admin
        - idir/cas_analyst
    - Add AppRole to bc_obps/registration/schema.py
    - Add GET endpoint for user role:  bc_obps/registration/api/user.py
    - Add my IDIR guid user to fixture: bc_obps/registration/fixtures/user.json

  

## 🚀 Impact:
   - App route protection implementation uses app directory to structure protected page routes based on NextAuth identity provider and user role as folders.
   - Each protected page has its own folder structure based on NextAuth JWT identity provider, NextAuth JWT user role, and a dashboard folder (used for dynamic breadcrumbs);   e.g. `app/(authenticated)/idir/cas_ admin/dashboard/operations/Page.tsx`
   - Access control logic is managed using Next.js middleware and NextAuth.js authentication JWT token.  The middleware intercepts requests, checks for a valid user session, and extracts user information from the JWT token. Based on JWT properties of identity_provider and role, the middleware  dynamically rewrites the request URL base request to the appropriate folder structure e.g. `http://localhost:3000/dashboard/operations` request maps to `http://localhost:3000/idir/cas_admin/dashboard/operations` for user `idir\cas_admin`
   - Using NextAuth JWT properties based static app directory to structure protected page routes and middleware logic to build dynamic responses allows for separation of concerns, maintainability and flexibility to delete or add roles without modifying existing access control logic.
 

## 📝 Notes:
- Using dynamic route creation based on JWT token properties attempts to align access control with the user's identity and role attributes which can then accommodate changes to user roles without requiring modifying existing access control logic. Static app directory protected routes can be added or removed without making any other changes. 
- One caveat is that page route components may be similar for multiple roles e.g. `app/(authenticated)/idir/cas_admin/dashboard/operations` and  `app/(authenticated)/idir/cas_analyst/dashboard/operations`; so, common components are used to share between protected page routes to reduce repetition.

## 🛠️ To Do:
- The BCEID tole is not finalized [321](https://app.zenhub.com/workspaces/team-moose-sprint-board-60d216175a2689000ec774d0/issues/gh/bcgov/cas-registration/321)
 - For outstanding FE PRs the legacy file structure remain but, new features will need to be integrated into protected routes as required.   [322](https://app.zenhub.com/workspaces/team-moose-sprint-board-60d216175a2689000ec774d0/issues/gh/bcgov/cas-registration/322)

##  🔬 Testing:

1.  From terminal command, start the app development server:
   ```
cd client && yarn dev
   ```
2. In your web browser, navigate to `http://localhost:3000`
3. You should be directed to the anonymous landing page
![image](https://github.com/bcgov/cas-registration/assets/120038448/bd8166b1-4f19-4993-82dc-0b3340d625e7)
4. Click "Program Administrator Log In"
5. You should be directed to IDIR login screen
![image](https://github.com/bcgov/cas-registration/assets/120038448/2e29ad50-559e-4c51-a756-c7953b9e17a8)
6. Login using your IDIR account
7. You should be redirected to the idir/cas_pending dashboard
![image](https://github.com/bcgov/cas-registration/assets/120038448/dd325db6-699a-4cb6-803b-20391e86b039)
8. Click your IDIR login name in the right hand header
9. Your user session information should display your GUID information
![image](https://github.com/bcgov/cas-registration/assets/120038448/bdaa40e1-2b14-47c8-818b-f1ce822f7a9a)
10. Add a user to `bc_obps/registration/fixtures/user.json` reflecting your IDIR GUID having `user_guid=your GUID information` and `app_role=cas_admin`
11.  Follow the [developer environment setup](https://github.com/bcgov/cas-registration/blob/develop/docs/developer-environment-setup.md#backend-environment-setup) to load fixtures  with your IDIR user data and start the API server
```
cd bc_obps
poetry shell
make loadfixtures
make run
   ```
12.  In your web browser, navigate to `http://localhost:3000`
13. Click "Log Out"
14. You should be directed to the anonymous landing page
15. Click "Program Administrator Log In"
16.  You should be directed to IDIR login screen
17. Login using your IDIR account
18. You should be redirected to the idir/cas_admin dashboard
![image](https://github.com/bcgov/cas-registration/assets/120038448/d6c62c5b-86e4-4f7b-8a08-1a03ee7bd27b)
19. Click your IDIR login name in the right hand header
20. Your user session information should display
![image](https://github.com/bcgov/cas-registration/assets/120038448/79f4457b-b592-49a0-8336-ba672537327a)
21. Your Role should reflect  `cas_admin` and your IP should be `idir`
22. Click the breadcrumb "Dashboard"
23. You should be redirected to the `idir/cas_admin` dashboard
24. Click all dashboard navigation tiles
25. You should be directed to the route page
26. Modify the browser URL to an app path accessible to `idir\cas_admin` e.g. `http://localhost:3000/idir/cas_admin/dashboard/operations`
27. You should be redirected to URL:  `http://localhost:3000/dashboard/operations` retaining the expected breadcrumb structure
28. Modify the browser URL to an app path **NOT** accessible to `idir\cas_admin` e.g.  `http://localhost:3000/idir/cas_analyst/dashboard/operations`
29. You should be redirected to "Page not found"
